File: Use.r

## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## poin tcloud to images
source("../src/pointcloud2images.R")

dir.create("jpg")
JPGfromPC(tree_pc = "../src/Original_Data/plot12_tree8.las",
          out_dir = "jpg")

## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## Predict using YOLOv5
## Clone Yolov5 and install dependencies
system("git clone https://github.com/ultralytics/yolov5")
system("cd yolov5 && pip install -r requirements.txt")

## Predict
command <- paste0("cd yolov5 && ",
                  "python3 detect.py ",
                  "--weights ../../src/whorl_detector_weights/best.pt ",
                  "--img 640 --conf 0.1 ",
                  "--source ../jpg ",
                  "--iou-thres=0.1 ",
                  "--line-thickness 1 ",
                  "--save-txt ",
                  "--save-conf")
system(command)

## move YOLOv5 predictios to jpg folder
file.copy(list.files("./yolov5/runs/detect", full.names = T),
          "./jpg", overwrite = TRUE, recursive = TRUE)

unlink("./yolov5/runs", recursive = TRUE)

## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## Parse from YOLOv5 format to Zmin Zmax and associated probability
source("../src/yoloPredictions2UTM.R")

parsePredictions(dir_predictions = "./jpg/exp/labels",
                 dir_orig_imgs = "./jpg")
                 
## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## postProcessingPredictions
source("../src/predictions_postProcessing.R")

postProcessingPredictions("./jpg/exp/labels", IOU = 0.001)

## @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
## Final data.
data.table::fread("Tree.csv")

File: Use.py
import os
import subprocess
from rpy2 import robjects
import shutil

jpg_folder = "jpg"

if not os.path.exists(jpg_folder):
   os.makedirs(jpg_folder)

robjects.r('''
source("src/pointcloud2images.R")

JPGfromPC(tree_pc = "src/Original_Data/plot12_tree8.las",
          out_dir = "jpg")
''')


if not os.path.exists("yolov5"):
   subprocess.run("git clone https://github.com/ultralytics/yolov5")

## Make and save predictions
## Reusable syntax for new models
from yolov5 import detect 

detect.run(
    weights="src/weights/best.pt",
    imgsz=(640,640),
    conf_thres=0.1,
    source="jpg",
    iou_thres=0.1,
    line_thickness=1,
    save_txt=True,
    save_conf=True
)

## move YOLOv5 predictios to jpg folder
robjects.r('''
        file.copy(list.files("yolov5/runs/detect", full.names = T), "jpg", overwrite = TRUE, recursive = TRUE)
          ''')

shutil.rmtree("yolov5/runs")

## Parse from YOLOv5 format to Zmin Zmax and associated probability
robjects.r('''
source("src/yoloPredictions2UTM.R")
parsePredictions(dir_predictions = "jpg/exp/labels",
                 dir_orig_imgs = "jpg")
''')

## postProcessingPredictions
robjects.r('''
source("src/predictions_postProcessing.R")
postProcessingPredictions("jpg/exp/labels", IOU = 0.001)
''')

## Check final data
import pandas as pd

print(pd.read_csv('Tree.csv'))